nixos/modules/services/backup/duplicity.nix

```diff
@@ -0,0 +1,99 @@
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+let cfg = config.services.duplicity;
+
+in {
+  options.services.duplicity = {
+    enable = mkEnableOption "backups with duplicity";
+
+    root = mkOption {
+      type = types.path;
+      default = "/";
+      description = ''
+        Root directory to backup.
+      '';
+    };
+
+    include = mkOption {
+      type = types.listOf types.path;
+      default = [];
+      example = [ "/home" ];
+      description = ''
+        List of paths to include into the backups. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for details on the syntax.
+      '';
+    };
+
+    exclude = mkOption {
+      type = types.listOf types.path;
+      default = [];
+      description = ''
+        List of paths to exclude from backups. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for details on the syntax.
+      '';
+    };
+
+    targetUrl = mkOption {
+      type = types.str;
+      example = "s3://host:port/prefix";
+      description = ''
+        Target url to backup to. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for supported urls.
+      '';
+    };
+
+    secretFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Path of a file containing secrets (gpg passphrase, access key...). It
+        is sourced before duplicity is run. For example, a file containing:
+          export PASSPHRASE='...'
+          export AWS_ACCESS_KEY_ID='...'
+          export AWS_SECRET_ACCESS_KEY='...'
+      '';
+    };
+
+    frequency = mkOption {
+      type = types.nullOr types.str;
+      default = "daily";
+      description = ''
+        Run duplicity with the given frequency (see
+        <citerefentry><refentrytitle>systemd.time</refentrytitle>
+        <manvolnum>7</manvolnum></citerefentry> for the format).
+        If null, do not run automatically.
+      '';
+    };
+
+    extraOptions = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      example = [ "--full-if-older-than 1M" ];
+      description = ''
+        Extra options to pass to duplicity.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.duplicity = {
+      description = "Backup files with duplicity.";
+      script =
+        ''
+          ${optionalString (cfg.secretFile != null) ''source "${toString cfg.secretFile}"''}
+          exec ${pkgs.duplicity}/bin/duplicity \
+            "${toString cfg.root}" \
+            "${toString cfg.targetUrl}" \
+            ${concatMapStringsSep " " (p: ''--include "${toString p}"'') cfg.include} \
+            ${concatMapStringsSep " " (p: ''--exclude "${toString p}"'') cfg.exclude} \
+            ${concatStringsSep " " cfg.extraOptions}
```

Comment:
Use `lib.escapeShellArgs` instead, maybe even for all arguments (by constructing a list with all of them, then using that function).